risc0/zkvm/src/host/client/exec.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    io::{BufReader, BufWriter},
+    path::Path,
+    process::{Command, Stdio},
+};
+
+use anyhow::{anyhow, Result};
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+
+use crate::ExecutorEnv;
+
+/// An event traced from the running VM.
+#[derive(Clone, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub enum TraceEvent {
+    /// An instruction has started at the given program counter
+    InstructionStart {
+        /// Cycle number since startup
+        cycle: u32,
+        /// Program counter of the instruction being executed
+        pc: u32,
+    },
+
+    /// A register has been set
+    RegisterSet {
+        /// Register ID (0-16)
+        reg: usize,
+        /// New value in the register
+        value: u32,
+    },
+
+    /// A memory location has been written
+    MemorySet {
+        /// Address of word that's been written
+        addr: u32,
+        /// Value of word that's been written
+        value: u32,
+    },
+}
+
+#[derive(Serialize, Deserialize)]
+enum PosixCmd {
+    ReadAvail,
+    Read,
+    Write,
+}
+
+#[derive(Serialize, Deserialize)]
+enum IpcSyscall {
```

Comment:
Calling these `Ipc*` seems to imply they can only be sent over IPC mechanisms, whereas is seems quite possible that this interaction might occur over a network in the future. Should we just call this `Syscall` and the others respectively? The module should disambiguate that this is not the same as a syscall elsewhere.